### ISSUE TITLE
mutex owner died -> POPO__CHUNK_LOCKING_ERROR

### ISSUE THREAD

**Issue by niclar** [eclipse-iceoryx/iceoryx#2193]
Hi, we just encountered a fatal crash in roudi (first time on b18411364d1cac72422c53a4b254fad0c9061204 since it came out), roudi logs below. 

Maybe an iceoryx "client" process died prior, but no iceoryx logging in them indicates this. And this should not crash roudi of course..
-Do you maybe have any pointers as to why we encountered this and how it should be remedied ?

(We're running 21 iceoryx client processes on the box, pinned and isolated, ~1500 publishers, and roudi on highest prio)

Thanks

## Required information

**Operating system:**
Ubuntu 22.04.2 LTS

**Compiler version:**
Clang 17.0.5

**Eclipse iceoryx version:**
b18411364d1cac72422c53a4b254fad0c9061204   (v2.0.x, 20231121)

**Observed result or behaviour:**

```
^[[0;90m2024-02-14 07:06:02.846 ^[[0;1;31;103m[Error]^[[m: /mnt/c/src/thirdparty/vcpkg/buildtrees/iceoryx/src/d0c9061204-f53d20e6d8.clean/iceoryx_hoofs/source/posix_wrapper/mutex.cpp:323 { expected<void, MutexLockError> iox::posix::mutex::lock() -> pthread_mutex_lock }  :::  [ 130 ]  Owner died
^[[0;90m2024-02-14 07:06:02.846 ^[[0;1;31;103m[Error]^[[m: The thread/process which owned the mutex died. The mutex is now in an inconsistent state and must be put into a consistent state again with Mutex::make_consistent()
^[[0;90m2024-02-14 07:06:02.846 ^[[0;1;97;41m[Fatal]^[[m: Locking of an inter-process mutex failed! This indicates that the application holding the lock was terminated or the resources were cleaned up by RouDi due to an unresponsive application.
^[[0;90m2024-02-14 07:06:02.846 ^[[0;1;31;103m[Error]^[[m: ICEORYX error! POPO__CHUNK_LOCKING_ERROR
libc++abi: terminating
```

**Comment by elBoberido** [comment]
@niclar this is related to #325. For the history feature there is a mutex involved. RouDi locks the mutex and can insert the samples into the queue without the interference of the publisher. This mutex is only contested when subscribers are added or removed but nevertheless, the publisher holds the lock for a short amount of time when publishing. If the application is terminated during this time and RouDi tries to clean up the resources of the process, it finds the corrupt mutex and terminates. I once tried to fix the problem but unfortunately, the data guarded by the mutex might also be corrupted and more refactoring would have been necessary to fully fix the problem. Other things came up and the issue never really happened in our setup.

This should actually only happen if either an application died or if RouDi assumed that an application died. But then you should see something like `Application foo not responding (last response 1509 milliseconds ago) --> removing it`. If the monitoring is turned off, then RouDi does not detect the dead application immediately but only when the application tries to re-register. But then there is also a log message. If you don't see any log messages for removing or re-registering an application, then I'm a bit at a loss and it could be anything, including [cosmic rays](https://www.bbc.com/future/article/20221011-how-space-weather-causes-computer-errors).

**Comment by niclar** [comment]
Looking at the logs again it looks like roudi incorrectly assumed 4 processes died prior. Don't know why it would assume this though. -Is there a starvation issue in the detection ?

**Comment by elBoberido** [comment]
@niclar this is related to #1361. I already improved the situation on master

**Comment by niclar** [comment]
-Looks like that fix is/was (2023-11-01) already in the code running (2023-11-23)

**Comment by elBoberido** [comment]
Oh, I just saw v2.0.x and thought it came from the release branch.

Is the system time changed abruptly so that jumps can occur?

Other scenarios that lead to this error might be when the threads which update the heartbeat value are not scheduled.

A workaround might be to massively increase the `PROCESS_KEEP_ALIVE_TIMEOUT`. Unfortunately it can currently not be changed via a cmake paremeter but that would not be difficult to add.

**Comment by niclar** [comment]
These thread(s) are part of the publisher(s) & not roudi right ? -I reckon they might have been starved in this instance.

**Comment by elBoberido** [comment]
It is a thread of the runtime. I think it should not be too difficult to add a `updateHeartbeat` method to the runtime. One could then call it manually.

**Comment by niclar** [comment]
This morning we encountered this fatal error again during (post) the morning restart routine of some iceoryx client processes.

**Comment by elBoberido** [comment]
Do you have more information, e.g. was the load during that time quite high?

**Comment by niclar** [comment]
cpu core utilization is low overall with roudi running on a dedicated core. we received this anew the other day seemingly "spontanously" a few hours after a mid day restart of ~20 roudi client processes.

**Comment by elBoberido** [comment]
Do you have some logs of spikes in the CPU utilization when the issue occurred?

**Comment by niclar** [comment]
@elfenpiff we do not. All pub/sub are instantiated with `historyCapacity=0` and `historyRequest=0` respectively.

**Comment by elBoberido** [comment]
It is not only the history. AFAIK, the mutex also guards the adding of the subscriber queue to the publisher.

Furthermore, the `POPO__CHUNK_LOCKING_ERROR` is only a symptom. The issue is that RouDi kills running applications because of the heartbeat thread not being scheduled. At least that's my assumption.

**Comment by elfenpiff** [comment]
@elBoberido 

> Furthermore, the `POPO__CHUNK_LOCKING_ERROR` is only a symptom. The issue is that RouDi kills running applications because of the heartbeat thread not being scheduled. At least that's my assumption.

This we can mitigate with turning monitoring off. But a better alternative would be to have a liveliness QoS that enforces some contracts like that a subscriber as to collect a sample latest after X seconds and the publisher must publisher every Y seconds a sample. But this would require some time-consuming refactoring.

> It is not only the history. AFAIK, the mutex also guards the adding of the subscriber queue to the publisher.

Yes, you are right. But this could be solved by bringing loffli into the play and a lock-free optional where the atomic signals that the thing is set or not. This is also some major refactoring but I think it is solvable.

**Comment by elBoberido** [comment]
@niclar did you encounter this problem again?

In the meantime I'm a bit hesitant on exposing the heartbeat triggering as public API. An alternative would be to increase the timeout for the detection of died processes. Currently it is about 1.5 seconds and the heartbeat should be sent every 300ms. If this happens again, can you check the RouDi log output for messages like `Application foo not responding (last response #### milliseconds ago) --> removing it` and report the amount of milliseconds?

**Comment by niclar** [comment]
As a matter of fact we did friday evening;

`^[[0;90m2024-04-12 18:35:37.238 ^[[0;1;93m[Warn ]^[[m: Application marketview.prd not responding (last response 1511 milliseconds ago) --> removing it
^[[0;90m2024-04-12 18:35:37.306 ^[[0;1;31;103m[Error]^[[m: /mnt/c/src/thirdparty/vcpkg/buildtrees/iceoryx/src/d0c9061204-bf50be8845.clean/iceoryx_hoofs/source/posix_wrapper/mutex.cpp:323 { expected<void, MutexLockError> iox::posix::mutex::lock() -> pthread_mutex_lock }  :::  [ 130 ]  Owner died
^[[0;90m2024-04-12 18:35:37.306 ^[[0;1;31;103m[Error]^[[m: The thread/process which owned the mutex died. The mutex is now in an inconsistent state and must be put into a consistent state again with Mutex::make_consistent()
^[[0;90m2024-04-12 18:35:37.306 ^[[0;1;97;41m[Fatal]^[[m: Locking of an inter-process mutex failed! This indicates that the application holding the lock was terminated or the resources were cleaned up by RouDi due to an unresponsive application.
^[[0;90m2024-04-12 18:35:37.306 ^[[0;1;31;103m[Error]^[[m: ICEORYX error! POPO__CHUNK_LOCKING_ERROR
libc++abi: terminating`

**Comment by elBoberido** [comment]
Okay, the deadline was missed only by a little bit. For some reasons the runtime thread which sends the heartbeat was not triggered for a long time. Is this on Windows or on Linux?

As a quick workaround I would suggest to try to change the `5` in this line in `iceoryx_posh_types.hpp` to `10`
```
constexpr units::Duration PROCESS_KEEP_ALIVE_TIMEOUT = 5 * PROCESS_KEEP_ALIVE_INTERVAL; // > PROCESS_KEEP_ALIVE_INTERVAL
```

It does not fix the problem but should make it much more unrealistic. You could increase the time even more. With e.g. `200` it would be one minute until RouDi would assume a process is dead. The 1.5 seconds came from a time where RouDi did not clean up the resources when a crashed application tried to re-register and therefore we kept it low in order to not have to wait to long. But now it could be indeed set to 1 minute which would not be too bad to clean the remainders of a crashed process up but should virtually prevent killing running applications.

The more I think about it, the more I come to the conclusion to just increase this timeout to eliminate this source for leaving a corrupt mutex behind.

**Comment by niclar** [comment]
Still encountering the issue

2024-04-22 11:32:58.571 [Warn ]: Application X not responding (last response 3046 milliseconds ago) --> removing it
2024-04-22 11:32:58.579 [Error]: /mnt/c/src/thirdparty/vcpkg/buildtrees/iceoryx/src/d0c9061204-2ee4ec5b88.clean/iceoryx_hoofs/source/posix_wrapper/mutex.cpp:323 { expected<void, MutexLockError> iox::posix::mutex::lock() -> pthread_mutex_lock }  :::  [ 130 ]  Owner died
2024-04-22 11:32:58.579 [Error]: The thread/process which owned the mutex died. The mutex is now in an inconsistent state and must be put into a consistent state again with Mutex::make_consistent()
2024-04-22 11:32:58.579 [Fatal]: Locking of an inter-process mutex failed! This indicates that the application holding the lock was terminated or the resources were cleaned up by RouDi due to an unresponsive application.
2024-04-22 11:32:58.579 [Error]: ICEORYX error! POPO__CHUNK_LOCKING_ERROR

**Comment by niclar** [comment]
We'll update to head in a day or two an see how that plays out...

**Comment by niclar** [comment]
It seems like HEAD might have solved the ipc queue  and this cleanup issue ,    let us run it a while longer, to be sure

**Comment by elBoberido** [comment]
Yeah, I somehow have the feeling that you still had a version which sent the heartbeat via the unix domain socket.

**Comment by niclar** [comment]
I managed to get this today again on f00cfe05c0048301bb72ade454771891696ee24c  (same settings as above/prior);

`^[[0;90m2024-04-23 11:44:43.001 ^[[0;1;93m[Warn ]^[[m: Application X not responding (last response 1550 milliseconds ago) --> removing it
^[[0;90m2024-04-23 11:44:43.006 ^[[0;1;31;103m[Error]^[[m: /mnt/c/src/thirdparty/vcpkg/buildtrees/iceoryx/src/91696ee24c-7813eef5f0.clean/iceoryx_hoofs/posix/sync/source/mutex.cpp:325 { expected<void, MutexLockError> iox::mutex::lock() -> iox_pthread_mutex_lock }  :::  [ 130 ]  Owner died
^[[0;90m2024-04-23 11:44:43.007 ^[[0;1;31;103m[Error]^[[m: The thread/process which owned the mutex died. The mutex is now in an inconsistent state and must be put into a consistent state again with Mutex::make_consistent()
^[[0;90m2024-04-23 11:44:43.007 ^[[0;1;97;41m[Fatal]^[[m: Locking of an inter-process mutex failed! This indicates that the application holding the lock was terminated or the resources were cleaned up by RouDi due to an unresponsive application.
^[[0;90m2024-04-23 11:44:43.007 ^[[0;1;97;41m[Fatal]^[[m: /mnt/c/src/thirdparty/vcpkg/buildtrees/iceoryx/src/91696ee24c-7813eef5f0.clean/iceoryx_posh/source/popo/building_blocks/locking_policy.cpp:42 [Fatal Error] [POPO__CHUNK_LOCKING_ERROR (code = 61)] in module [iceoryx_posh (id = 2)]
^[[0;90m2024-04-23 11:44:43.007 ^[[0;1;97;41m[Fatal]^[[m: /mnt/c/src/thirdparty/vcpkg/buildtrees/iceoryx/src/91696ee24c-7813eef5f0.clean/iceoryx_posh/source/popo/building_blocks/locking_policy.cpp:42 [PANIC]`

**Comment by niclar** [comment]
I'll try setting the timeout to a minute

**Comment by niclar** [comment]
we still encounter it with the 1min delay;

2024-04-23 16:10:35.281 [Warn ]: Application X not responding (last response 1501 milliseconds ago) --> removing it
2024-04-23 16:10:35.286 [Error]: /mnt/c/src/thirdparty/vcpkg/buildtrees/iceoryx/src/91696ee24c-11145aa309.clean/iceoryx_hoofs/posix/sync/source/mutex.cpp:325 { expected<void, MutexLockError> iox::mutex::lock() -> iox_pthread_mutex_lock }  :::  [ 130 ]  Owner died
2024-04-23 16:10:35.286 [Error]: The thread/process which owned the mutex died. The mutex is now in an inconsistent state and must be put into a consistent state again with Mutex::make_consistent()
2024-04-23 16:10:35.286 [Fatal]: Locking of an inter-process mutex failed! This indicates that the application holding the lock was terminated or the resources were cleaned up by RouDi due to an unresponsive application.
2024-04-23 16:10:35.286 [Fatal]: /mnt/c/src/thirdparty/vcpkg/buildtrees/iceoryx/src/91696ee24c-11145aa309.clean/iceoryx_posh/source/popo/building_blocks/locking_policy.cpp:42 [Fatal Error] [POPO__CHUNK_LOCKING_ERROR (code = 61)] in module [iceoryx_posh (id = 2)]
2024-04-23 16:10:35.286 [Fatal]: /mnt/c/src/thirdparty/vcpkg/buildtrees/iceoryx/src/91696ee24c-11145aa309.clean/iceoryx_posh/source/popo/building_blocks/locking_policy.cpp:42 [PANIC]

**Comment by elBoberido** [comment]
The log does not fit the 1minute timeout. It still terminates the app after 1.5 seconds.

**Comment by niclar** [comment]
1 minute timeout is set/compiled however. No idea why this happens..

**Comment by elBoberido** [comment]
Can you set it to `60_s`?

**Comment by elBoberido** [comment]
Oh, since you are using the main branch, can you also run RouDi with `-x commitId` to ensure that all applications use the same iceoryx version. If you want to be more strict you could also deactivate reproducible builds and use `-x buildDate`. The default is `patch` which allows all applications with the same patch level to connect to RouDi.

**Comment by PMSRGroup** [comment]
@niclar Have you managed to solve the issue? If so, how? We are getting the same exact problem, currently trying the options listed here but no luck so far. Will update to latest later tonight and let it run for a few days.
Currently running: f602986bc77fa9b73aefa9a86dc0af15fa754700

**Comment by rbdm-qnt** [comment]
Having the same issue here aswell, on commit f6dc4e1 (latest). @niclar @elBoberido any way to make it less frequent or solve it completely?

**Comment by elBoberido** [comment]
@rbdm-qnt if the monitoring is turned off, then this will not happen but in the case the application crashes, the resources are only cleaned up when RouDi shuts down or when the application re-registers.

Can you try the following changes in `iceoryx_posh_types.hpp`
```diff
-- constexpr units::Duration PROCESS_KEEP_ALIVE_TIMEOUT = 5 * PROCESS_KEEP_ALIVE_INTERVAL;
++ constexpr units::Duration PROCESS_KEEP_ALIVE_TIMEOUT = 60_s;
```

My guess is that this issue is related to jumps in the system time like mentioned further above. I was just able to trigger the issue by changing the system time with `timedatectl`. I think the problem is that we rely on the timestamp from the last heartbeat of the application. If there are jumps in the system time, the heartbeat thread of the application as well as the monitoring thread of RouDi will be woken up and now there is a race between the application sending the heartbeat with the updated timestamp and RouDi checking for the liveliness of the applications. If RouDi wins and the jump was into the future, it assumes that the application did not send the heartbeat for too long and therefore is in a degraded state.

Please check if your system time jumps and if increasing the `PROCESS_KEEP_ALIVE_TIMEOUT` timeout reduces the frequency.

The proper solution would probably be to use a counter instead of a timestamp and then we could have a heuristic like when the counter was not incremented for 5 consecutive times RouDi checks for the liveliness, the application can be considered to be in a degraded state.

**Comment by elBoberido** [comment]
@rbdm-qnt @PMSRGroup @niclar did you have a chance to check whether my theory with the jumps in system time  is correct and the workaround works?

**Comment by passchaos** [comment]
@elBoberido Can the timestamp used by the heartbeat be the system boot time, so that it is not affected by the system time

**Comment by elBoberido** [comment]
@passchaos are you also experiencing this issue?

I just had a look at the code and we are already using `CLOCK_MONOTONIC`, which is related to the boot time on linux. Unfortunately this is also affected by adjustments performed by `adjtime` and `NTP`. There is `CLOCK_MONOTONIC_RAW` which should not be affected but I can also trigger the error with that clock source by performing these steps.

```
sudo timedatectl set-ntp false
sudo timedatectl set-time 18:30:00
sudo timedatectl set-true
```

At least in one of three runs, RouDi assumes the application did not send the heartbeat.

The workaround would be to increase the `PROCESS_KEEP_ALIVE_TIMEOUT` to a value larger than the jumps in the clock.

A proper fix would be to rework the heartbeat handling and I also have an idea how to do it but I'm currently quite busy with other topics.

**Comment by elBoberido** [comment]
@passchaos the code in `duration.cpp` relates only to conversions from `Duration` to `timespec` and is not related to the heartbeat.

The heartbeat uses `CLOCK_MONOTONIC`
https://github.com/eclipse-iceoryx/iceoryx/blob/main/iceoryx_posh/source/runtime/heartbeat.cpp#L50-L60

**Comment by passchaos** [comment]
@elBoberido Well, I know this is the logic of the heartbeat, and each process involved in the communication sends a Keepalive signal to roudi, but because each process uses a timer to send a Keepalive signal, the timer uses a time source that is not monotonic (see my previous link). 
Therefore, if the system time is set back to the previous time, the timer considers that the time is not up and does not send a Keepalive signal to roudi. 
When I change to monotonic, the loss of heartbeat due to time adjustment no longer occurs.

**Comment by elBoberido** [comment]
@passchaos ah, right. The thread is waits on a semaphore and that causes the issue. The problem is that `sem_timedwait` requires `CLOCK_REALTIME`, see also https://www.man7.org/linux/man-pages/man3/sem_timedwait.3p.html.

Have you checked whether on of your cores runs at 100% after the change? I assume that the heartbeat thread does not sleep anymore and returns immediately from the `sem_timedwait` call.

**Comment by passchaos** [comment]
@elBoberido Yeah, I have also noticed this problem, it is a sad thing

**Comment by hrudhansh** [comment]
This issue is showing for me 7/10 times upon stopping publisher process. 

How to recreate?
1. Run a pub-sub with no delay in between publishes.
2. Register a SIGINT signal handler in your main like [`signal(SIGINT, SignalHandler);`](https://www.tutorialspoint.com/cplusplus/cpp_signal_handling.htm)
3. Upon ctrl+c on the pub process, you would see it stall. 
4. Upon ctrl+x again you would see the above error in iox-roudi. 

Hopefully this helps you recreate/debug this on your end. 

On my end, I ran gdb on the pub process with  `-exec handle SIGINT nostop` & `-exec handle SIGINT pass`, a breakpoint on the `exit(sigint);` and called `pkill -SIGINT publisher` in a separate terminal. I noticed: 
- In cases where it fails, the signal handler seems to be called in the middle of a publish thread. 
- In both cases (fail vs no-fail), there is a KeepAlive thread running with the semaphore. So not sure if it's this. 

**How do I stop/interrupt the publisher thread gracefully?** It seems like just letting the de-constructor call is not enough.
**How do I stop the KeepAlive thread gracefully?** 

Not advocating a quick fix, but I wonder if we can put an atomic bool somewhere in the publish thread that we can call to abruptly stop what it's doing, and tear down. 

Note: This test was done on commit: b2cd72bdc789bcf7601cb112c6078c47d533d798 which is not latest. But the same issue persists on latest.

**Comment by elBoberido** [comment]
@hrudhansh I think the root of your problem is a different one than from this issue. It just manifests in the same error code. Can you please create a separate issue for your problem?

**Comment by elBoberido** [comment]
@niclar @PMSRGroup @rbdm-qnt @passchaos shall we schedule a community meeting to discuss the problem, solution approaches and how to distribute the work?

**Comment by gpalmer-latai** [comment]
Maybe I'm interested as well. I haven't enabled monitoring but I encounter probably the same error case when shutting down RouDi - likely because some process has crashed at some point while holding the mutex.

**Comment by niclar** [comment]
yes please do, it's a serious bug that needs to be addressed

**Comment by gpalmer-latai** [comment]
What are folks thoughts on the idea of a short-term mitigation whereby this error becomes "non-fatal" and is treated as a leaked resource? Basically it means that this publisher port is in a corrupted state and may not be reused until the whole system (RouDI) is restarted. The default configuration has an allotment for 512 publisher ports overall so in practice this seems like a manageable scenario, so long as you aren't trying to keep restarting the failing process and repeatedly corrupting publisher ports.

**Comment by gpalmer-latai** [comment]
That works for me.

**Comment by niclar** [comment]
@elBoberido meetup at Tuesday at 5 p.m. CEST works fine.
@gpalmer-latai at least for us such a stop gap solution with proper warnings (#free ports #currupt ports) would suffice.

**Comment by hrudhansh** [comment]
Interested in the meeting as well, will possibly enlighten me a fix for #2304

**Comment by elBoberido** [comment]
https://github.com/eclipse-iceoryx/iceoryx/wiki/2024-06-25-Eclipse-iceoryx-developer-meetup

**Comment by NeilZhy** [comment]
Is there any progress in fixing this issue?

**Comment by elBoberido** [comment]
@NeilZhy unfortunately this is not trivial to fix and requires quite some time. Up until now, there was no-one who took the time to properly fix it or was willing to fund it since there are also workarounds that kind of work.

The core team is currently focused on iceoryx2 where this problem does not occur.

**Comment by rbdm-qnt** [comment]
@elBoberido sorry for missing the community meetup, just saw a notification mail. This error still keeps crashing our programs at least twice a week, we had to make a restart mechanism to mitigate it. Is iceoryx2 ready to use in C++ or is it still Rust only? When can we expect to be able to switch? Thanks

**Comment by elfenpiff** [comment]
@rbdm-qnt The C++ API is nearly feature complete (only some new small iceoryx2 features are missing, which are not required when you port from classic iceoryx). The last remaining feature that is missing in iceoryx2 to reach feature parity with iceoryx is request response.
We started working on it and intend to implement this with C, C++, and Rust API latest in Q1 2025.

**Comment by elBoberido** [comment]
@lucabart97 welcome back :)

Your proposal would work. The problem is how to ping the other app without causing issues. When RouDi shuts down, it sends a SIGTERM to gracefully shut down all registered applications and then another signal to check if the app is still running and if it did not terminate in a graceful manner after some time a SIGKILL. We cannot do this while the app ist still operational and RouDi is not shutting down.

What might work is to just have a counter as heartbeat instead of a timestamp. In the `ProcessManager`, there could be a local buffer which stores the value of the counter and the number calls to `monitoringProcesses` since the counter did not change. This should eliminate at least most of the reasons which trigger this bug.

Other workarounds would be to either increase the timeout or deactivate the monitoring. If you can detect that an application crashed and restart it, the cleanup will be done when the application re-registers at RouDi.

I would also suggest to port to iceoryx2, which does not have this issue and is already more feature-full and more robust.

**Comment by lucabart97** [comment]
Thank you for your feedback, we saw in iceoryx2 that the c/c++ code was in beta, but now it is done, right?

> Other workarounds would be to either increase the timeout or deactivate the monitoring. If you can detect that an application crashed and restart it, the cleanup will be done when the application re-registers at RouDi.
Yes, but the application name must be the same, right? 
We use an application name plus 4 random values to avoid interference between nodes.

**Comment by elBoberido** [comment]
> Thank you for your feedback, we saw in iceoryx2 that the c/c++ code was in beta, but now it is done, right?

Yes, with the v0.6 the C++ bindings were finished. A few days ago, the bindings were even made thread-safe by default on main. The next release will be done once the Python bindings are finished, which will probably be in a few weeks.

> Yes, but the application name must be the same, right? We use an application name plus 4 random values to avoid interference between nodes.

I see. In this case `You're holding it wrong`™. Joking aside, the idea was that the runtime name should prevent to run the exact same application twice and if required, to use a config file or cmd line argument to add additional semantic meaning to the runtime nane, e.g. `*-front` and `*-rear`. Of course we might have overlooked use cases where this is not easily possible.

Anyway, in iceoryx2, the node does not need to have a name, but it is recommended to help debugging.
